File: web/src/lib/api/schema/announcement.ts
/**
 * Schema for the `GET /api/announcements/active` response.
 *
 * Mirrors the `AnnouncementResponse` DTO emitted by yog-api: one
 * operator announcement per entry, most severe first then most recent
 * (the ordering is the API contract — the client just takes the head).
 *
 * Notes:
 *   - `kind` and `severity` are closed enums (CHECK-constrained in
 *     the DB) — `z.enum` surfaces drift as a validation failure.
 *   - `severity` is the *announcement* scale, deliberately distinct
 *     from the signal severity concept even though the tags coincide
 *     today (editorial display choice vs detector conclusion).
 *   - `linkUrl` is a plain string, not `z.url()`: the main use case is
 *     a relative in-app target like `/changelog#v0.1.1`.
 */

import * as z from "zod";

import { Rfc3339 } from "./shared";

/** What an announcement is about — drives the label chip. */
export const AnnouncementKindSchema = z.enum([
  "maintenance",
  "incident",
  "release",
  "beta",
]);

/** Announcement kind — `"maintenance" | "incident" | "release" | "beta"`. */
export type AnnouncementKind = z.infer<typeof AnnouncementKindSchema>;

/** How prominently to display it — drives the banner styling. */
export const AnnouncementSeveritySchema = z.enum([
  "info",
  "warning",
  "critical",
]);

/** Announcement severity — `"info" | "warning" | "critical"`. */
export type AnnouncementSeverity = z.infer<typeof AnnouncementSeveritySchema>;

/** Schema for one active announcement. */
export const AnnouncementSchema = z.object({
  // Storage identity — the dismiss-cookie key.
  id: z.number(),
  kind: AnnouncementKindSchema,
  severity: AnnouncementSeveritySchema,
  // Free operator text (English, v1 decision).
  message: z.string(),
  // Optional target; usually a relative path (see file header).
  linkUrl: z.string().nullable(),
  startsAt: Rfc3339,
  // null = open-ended (shown until the operator closes it).
  endsAt: Rfc3339.nullable(),
});

/** Validated announcement payload. */
export type AnnouncementResponse = z.infer<typeof AnnouncementSchema>;

/** The endpoint returns the full active set as a bare array. */
export const AnnouncementListSchema = z.array(AnnouncementSchema);

File: crates/api/src/http/dto/response/announcement.rs
//! Response DTO for `GET /api/announcements/active`.
//!
//! Kept separate from the domain type, like every other `*Response` in
//! this module — the domain model never leaks into the JSON wire shape.

use chrono::{DateTime, Utc};
use serde::Serialize;

use yog_core::domain::Announcement;

/// One active operator announcement, as displayed by the web banner.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct AnnouncementResponse {
    /// Storage identity — the client's dismiss-cookie key.
    id: i64,

    /// What it is about: "maintenance" | "incident" | "release" | "beta".
    kind: String,

    /// Display prominence: "info" | "warning" | "critical".
    severity: String,

    /// Free operator text (English, v1 decision).
    message: String,

    /// Optional target, e.g. `/changelog#v0.1.1`. `null` when absent.
    link_url: Option<String>,

    /// Start of the display window.
    starts_at: DateTime<Utc>,

    /// End of the display window; `null` = open-ended.
    ends_at: Option<DateTime<Utc>>,
}

impl From<Announcement> for AnnouncementResponse {
    fn from(a: Announcement) -> Self {
        Self {
            id: a.id,
            kind: a.kind.as_str().to_string(),
            severity: a.severity.as_str().to_string(),
            message: a.message,
            link_url: a.link_url,
            starts_at: a.starts_at,
            ends_at: a.ends_at,
        }
    }
}
